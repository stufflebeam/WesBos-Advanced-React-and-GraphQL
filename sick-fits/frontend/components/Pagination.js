import Head from 'next/head';
import Link from 'next/link';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faLongArrowAltLeft,
  faLongArrowAltRight,
} from '@fortawesome/free-solid-svg-icons';
import PaginationStyles from './styles/PaginationStyles';
import LoadingAnimation from './LoadingAnimation';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allProductsMeta {
      count
    }
  }
`;

export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);
  if (loading) return <LoadingAnimation config={{ color: '#FF0000' }} />;
  if (error) return <DisplayError error={error} />;
  console.log('[Pagination]: data:', data);
  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles>
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`}>
        <a aria-disabled={page <= 1}>
          <FontAwesomeIcon icon={faLongArrowAltLeft} className="fa-icon" />
          Prev
        </a>
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>
          Next
          <FontAwesomeIcon icon={faLongArrowAltRight} className="fa-icon" />
        </a>
      </Link>
    </PaginationStyles>
  );
}
