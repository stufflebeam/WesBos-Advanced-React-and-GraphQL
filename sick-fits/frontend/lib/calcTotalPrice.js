export default function calcTotalPrice(cart) {
  //   console.log('[calcTotalPrice] cart', cart);
  if (!cart) return 0;
  return cart.reduce((tally, cartItem) => {
    // We run the following check because it is possible for an item to
    // be deleted while it is still in a user's cart.
    if (!cartItem.product) return tally;
    return tally + cartItem.product.price * cartItem.quantity;
  }, 0);
}
