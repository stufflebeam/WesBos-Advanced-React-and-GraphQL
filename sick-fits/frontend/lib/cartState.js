import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider
  // We will store data (state) and functionality (updaters) here
  // and anyone can access this data and the updaters via the consumer
  //   const cartOpen = true;
  const [cartOpen, setCartOpen] = useState(true);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, openCart, closeCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// This is a custom hook for accessing the data in our provider (the cart local state)
function useCart() {
  // we use a consumer here, to access the local state
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, LocalStateContext, useCart };
