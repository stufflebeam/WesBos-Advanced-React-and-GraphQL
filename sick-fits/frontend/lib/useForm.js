import { useEffect, useState } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our form inputs
  const [inputs, setInputs] = useState(initial);
  // const initialValues = Object.values(initial).join('');
  // This 'feels' better than the suggested solution (above), as it avoids useEffect()
  // being called due to object property order not being guaranteed.
  // This way, it only gets called when the form is first populated with fetched data
  // and the values go from '' to something else.
  const initialValuesEmpty = Object.values(initial).join('') === '';
  // console.log('[useForm] initialValuesEmpty: ', initialValuesEmpty);

  useEffect(() => {
    // This function runs when the things we are watching change
    // console.log(
    //   '[useForm] useEffect() initialValuesEmpty: ',
    //   initialValuesEmpty
    // );
    setInputs(initial);
  }, [initialValuesEmpty]);

  // create a function that updates the state object
  const handleChange = (e) => {
    let { value, name, type } = e.target;

    // check to see if the input is a number, and if it is, convert it to a number
    if (type === 'number') {
      value = parseInt(value);
    }

    // check to see if the input is a file and handle it accordingly if so
    if (type === 'file') {
      [value] = e.target.files;
    }

    // console.log('[handleChange]', { value, name, type });
    setInputs({
      // copy the existing state first
      ...inputs,
      // update the state with the new input value
      [name]: value,
    });
  };

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    // setInputs({}); // doesn't work
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  // return the things we wan to surface from this custom hook
  // specifically, an object that includes the state and a function to update it
  return { inputs, handleChange, resetForm, clearForm };
}
