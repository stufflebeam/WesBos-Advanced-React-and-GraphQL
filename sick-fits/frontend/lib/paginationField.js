// import {PAGINATION_QUERY} from '../graphql/queries'; // TODO: move all queries to graphql folder
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // This tells Apollo that we will take care of everything.
    read(existing = [], { args, cache }) {
      //   console.log({ existing, args, cache });
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      //   console.log({ first, skip, data });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1; // Math.ceil(skip / first) + 1;
      const pageCount = Math.ceil(count / first);
      console.log('[paginationField] read():', {
        skip,
        first,
        count,
        page,
        pageCount,
      });

      // Check to see if we have the requested items in the cache.
      const items = existing.slice(skip, skip + first).filter((x) => x); // TODO: Should this be .filter((x) => !!x);

      /* 
      IF
        There are items
        AND the number of items is not enough to satisfy the request (i.e. there are not enough to fill the page)
        AND we are on the last page
      THEN just send the data back to Apollo
      */

      // TODO: Handle the case where a user tries to navigate to a page that doesn't exist.
      //       For example, if there are 10 items total, and we're showing 4 items per page,
      //       and the user navigates to page 5, we should either route them to the main
      //       products page OR route them to the last products page by updating the URL
      //       from products/5 to products/3 and showing them the last 2 items.

      if (items.length && items.length !== first && page === pageCount) {
        return items;
      }

      if (items.length !== first) {
        // If we don't have all the items, we need to fetch them.
        return false;
      }

      // If we have all the items, we can just return them.
      if (items.length) {
        console.log(
          `[paginationField]: There are ${items.length} items in the cache. Sending them to Apollo...`
        );
        return items;
      }

      // If either of the previous return statements didn't return anything, we need to fetch the items,
      // so we return false.
      return false;

      // When we run a query, the first thing that Apollo does is ask the read() function for those items.
      // When this happens, we can do one of two things:
      // 1. If we have the data, we can return it.
      // 2. If we don't have the data, we can return null or false, which will tell Apollo to run the query via the network.
    },
    merge(existing, incoming, { args, cache }) {
      // When Apollo gets query results back from the network for items that the read() function returned false for,
      // it will call the merge() function to merge the results into the cache.
      // So, this is where we decide how we want to put the returned data into the cache.
      const { skip, first } = args;

      console.log('[paginationField]: merging items from the network...]', {
        existing,
        incoming,
        args,
        cache,
      });
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log('[paginationField] merge():', { merged });

      // return the merged items
      // at this point, Apollo will call the read() function again to see if it can return the
      // merged items from the cache.
      return merged;
    },
  };
}
