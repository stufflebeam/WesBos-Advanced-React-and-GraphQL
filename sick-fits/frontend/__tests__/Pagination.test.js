import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { CartStateProvider } from '../lib/cartState';
import Pagination from '../components/Pagination';
import { makePaginationMocksFor } from '../lib/testUtils';
import { perPage } from '../config';

// NOTE: All of the tests here that have numbers in them are based on a config
//       of perPage = 2. If you change the perPage value in config.js, you'll
//       need to update these tests.
//       TODO: Make these tests dynamic based on the perPage value in config.js

describe('<Pagination />', () => {
  it('renders the loader during loading', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(1)}>
          <Pagination page={1} />
        </MockedProvider>
      </CartStateProvider>
    );
    // debug(container);
  });

  it("renders correctly when there's one page", async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(1)}>
          <Pagination page={1} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);
    expect(container).toHaveTextContent('Page 1 of 1');
    expect(container).toHaveTextContent('1 Item Total');

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/0');
    expect(prevLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/2');
    expect(nextLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('renders correctly when on page one of two', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(4)}>
          <Pagination page={1} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);
    expect(container).toHaveTextContent('Page 1 of 2');
    expect(container).toHaveTextContent('4 Items Total');

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/0');
    expect(prevLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/2');
    expect(nextLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('renders correctly when on page two of two', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(4)}>
          <Pagination page={2} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);
    expect(container).toHaveTextContent('Page 2 of 2');
    expect(container).toHaveTextContent('4 Items Total');

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/1');
    expect(prevLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/3');
    expect(nextLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('renders correctly for 15 items, when on page 3 of 8', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(15)}>
          <Pagination page={3} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);
    expect(container).toHaveTextContent('Page 3 of 8');
    expect(container).toHaveTextContent('15 Items Total');

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/2');
    expect(prevLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/4');
    expect(nextLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('disables the prev button and enables the next button on the first page', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(6)}>
          <Pagination page={1} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/0');
    expect(prevLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/2');
    expect(nextLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('enables the prev button and disables the next button on the last page', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(6)}>
          <Pagination page={3} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/2');
    expect(prevLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/4');
    expect(nextLink).toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('enables the prev and next buttons on the middle pages', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(6)}>
          <Pagination page={2} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);

    expect(container).toHaveTextContent('Prev');
    const prevLink = screen.getByText('Prev');
    expect(prevLink).toBeInTheDocument();
    expect(prevLink).toHaveAttribute('href', '/products/1');
    expect(prevLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toHaveTextContent('Next');
    const nextLink = screen.getByText('Next');
    expect(nextLink).toBeInTheDocument();
    expect(nextLink).toHaveAttribute('href', '/products/3');
    expect(nextLink).not.toHaveAttribute('aria-disabled', 'true');

    expect(container).toMatchSnapshot();
  });

  it('enables the prev and next buttons on the middle pages [alternate test setup]', async () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider mocks={makePaginationMocksFor(6)}>
          <Pagination page={2} />
        </MockedProvider>
      </CartStateProvider>
    );
    // await screen.findByText('Prev');
    await screen.findByTestId('pagination');
    // debug(container);

    const prevButton = screen.getByText(/Prev/);
    const nextButton = screen.getByText(/Next/);
    expect(prevButton).toHaveAttribute('href', '/products/1');
    expect(prevButton).toHaveAttribute('aria-disabled', 'false');
    expect(nextButton).toHaveAttribute('href', '/products/3');
    expect(nextButton).toHaveAttribute('aria-disabled', 'false');

    expect(container).toMatchSnapshot();
  });
});
