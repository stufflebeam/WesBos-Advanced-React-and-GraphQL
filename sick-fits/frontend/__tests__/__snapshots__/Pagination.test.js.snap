// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> disables the prev button and enables the next button on the first page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      1
       of 
      3
    </p>
    <p>
      6
       Item
      s
       Total
    </p>
    <a
      aria-disabled="false"
      href="/products/2"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> enables the prev and next buttons on the middle pages [alternate test setup] 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/1"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      2
       of 
      3
    </p>
    <p>
      6
       Item
      s
       Total
    </p>
    <a
      aria-disabled="false"
      href="/products/3"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> enables the prev and next buttons on the middle pages 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/1"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      2
       of 
      3
    </p>
    <p>
      6
       Item
      s
       Total
    </p>
    <a
      aria-disabled="false"
      href="/products/3"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> enables the prev button and disables the next button on the last page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/2"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      3
       of 
      3
    </p>
    <p>
      6
       Item
      s
       Total
    </p>
    <a
      aria-disabled="true"
      href="/products/4"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> renders correctly for 15 items, when on page 3 of 8 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/2"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      3
       of 
      8
    </p>
    <p>
      15
       Item
      s
       Total
    </p>
    <a
      aria-disabled="false"
      href="/products/4"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> renders correctly when on page one of two 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      1
       of 
      2
    </p>
    <p>
      4
       Item
      s
       Total
    </p>
    <a
      aria-disabled="false"
      href="/products/2"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> renders correctly when on page two of two 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="false"
      href="/products/1"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      2
       of 
      2
    </p>
    <p>
      4
       Item
      s
       Total
    </p>
    <a
      aria-disabled="true"
      href="/products/3"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`<Pagination /> renders correctly when there's one page 1`] = `
<div>
  <div
    class="PaginationStyles-sc-aduuar-0 cnxjRn"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      href="/products/0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-left-long fa-icon"
        data-icon="left-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 273.7 497.7 288 480 288H160.1l0 72c0 9.547-5.66 18.19-14.42 22c-8.754 3.812-18.95 2.077-25.94-4.407l-112.1-104c-10.24-9.5-10.24-25.69 0-35.19l112.1-104c6.992-6.484 17.18-8.218 25.94-4.406C154.4 133.8 160.1 142.5 160.1 151.1L160.1 224H480C497.7 224 512 238.3 512 256z"
          fill="currentColor"
        />
      </svg>
      Prev
    </a>
    <p>
      Page 
      1
       of 
      1
    </p>
    <p>
      1
       Item
      
       Total
    </p>
    <a
      aria-disabled="true"
      href="/products/2"
    >
      Next
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-long fa-icon"
        data-icon="right-long"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504.3 273.6l-112.1 104c-6.992 6.484-17.18 8.218-25.94 4.406c-8.758-3.812-14.42-12.45-14.42-21.1L351.9 288H32C14.33 288 .0002 273.7 .0002 255.1S14.33 224 32 224h319.9l0-72c0-9.547 5.66-18.19 14.42-22c8.754-3.809 18.95-2.075 25.94 4.41l112.1 104C514.6 247.9 514.6 264.1 504.3 273.6z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</div>
`;
